Backport of:

From 75cecc4e39ad28f989700e261445adde0c88dc86 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bastien=20ROUCARI=C3=88S?= <roucaries.bastien@gmail.com>
Date: Mon, 29 Dec 2014 10:51:48 +0100
Subject: [PATCH] Fix correctly the xpm crash problem

The xpm crash was fixed but the xpm result was not correct (xpm file was destroyed).

Fix correctly.

Bug-Debian: https://bugs.debian.org/773980
Origin: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=773980#47 and http://www.imagemagick.org/discourse-server/viewtopic.php?f=3&t=26787&p=118367#p118368
---
 coders/xpm.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

Index: imagemagick-6.6.9.7/coders/xpm.c
===================================================================
--- imagemagick-6.6.9.7.orig/coders/xpm.c	2016-11-15 13:45:14.518695146 -0500
+++ imagemagick-6.6.9.7/coders/xpm.c	2016-11-15 13:45:14.514695097 -0500
@@ -157,7 +157,8 @@
   p=source;
   while (length-- && (*p != '\0'))
     *destination++=(*p++);
-  *destination='\0';
+  if (length != 0)
+    *destination='\0';
   return((size_t) (p-source));
 }
 
@@ -359,7 +360,7 @@
   {
     p=next;
     next=NextXPMLine(p);
-    (void) CopyXPMColor(key,p,MagickMin((size_t) width,MaxTextExtent));
+    (void) CopyXPMColor(key,p,MagickMin((size_t) width,MaxTextExtent-1));
     status=AddValueToSplayTree(xpm_colors,ConstantString(key),(void *) j);
     /*
       Parse color.
@@ -374,7 +375,7 @@
           break;
         if (next != (char *) NULL)
           (void) CopyXPMColor(target,q,MagickMin((size_t) (next-q),
-            MaxTextExtent));
+            MaxTextExtent-1));
         else
           (void) CopyMagickString(target,q,MaxTextExtent);
         q=ParseXPMColor(target);
@@ -411,7 +412,7 @@
         indexes=GetAuthenticIndexQueue(image);
         for (x=0; x < (ssize_t) image->columns; x++)
         {
-          p+=CopyXPMColor(key,p,MagickMin(width,MaxTextExtent));
+          p+=CopyXPMColor(key,p,MagickMin(width,MaxTextExtent-1));
           j=(ssize_t) GetValueFromSplayTree(xpm_colors,key);
           if (image->storage_class == PseudoClass)
             SetIndexPixelComponent(indexes+x,j);
